packages/hydrogen-react/src/getProductOptions.ts

```diff
@@ -0,0 +1,442 @@
+
+  // Check for nested options requirements
+  if (product.options) {
+    const firstOption = product?.options[0];
+
+    if (checkAll && !firstOption?.name) {
+      validParam = logErrorAndReturnFalse('options.name');
+    }
+
+    // Check for options.optionValues
+    if (product?.options[0]?.optionValues) {
+      const firstOptionValues = product.options[0].optionValues[0];
+
+      // Check for options.optionValues.name
+      if (checkAll && !firstOptionValues?.name) {
+        validParam = logErrorAndReturnFalse('options.optionValues.name');
+      }
+
+      // Check for options.optionValues.firstSelectableVariant
+      if (firstOptionValues?.firstSelectableVariant) {
+        // check product variant
+        validParam = checkProductVariantParam(
+          firstOptionValues.firstSelectableVariant,
+          'options.optionValues.firstSelectableVariant',
+          validParam,
+          checkAll,
+        );
+      } else {
+        validParam = logErrorAndReturnFalse(
+          'options.optionValues.firstSelectableVariant',
+        );
+      }
+    } else {
+      validParam = logErrorAndReturnFalse('options.optionValues');
+    }
+  }
+
+  // Check for nested selectedOrFirstAvailableVariant requirements
+  if (product.selectedOrFirstAvailableVariant) {
+    validParam = checkProductVariantParam(
+      product.selectedOrFirstAvailableVariant,
+      'selectedOrFirstAvailableVariant',
+      validParam,
+      checkAll,
+    );
+  }
+
+  // Check for nested adjacentVariants requirements
+  if (!!product.adjacentVariants && product.adjacentVariants[0]) {
+    validParam = checkProductVariantParam(
+      product.adjacentVariants[0],
+      'adjacentVariants',
+      validParam,
+      checkAll,
+    );
+  }
+
+  return (validParam ? product : {}) as Product;
+}
+
+function checkProductVariantParam(
+  variant: RecursivePartial<ProductVariant>,
+  key: string,
+  currentValidParamState: boolean,
+  checkAll: boolean,
+): boolean {
+  let validParam = currentValidParamState;
+
+  if (checkAll && !variant.product?.handle) {
+    validParam = logErrorAndReturnFalse(`${key}.product.handle`);
+  }
+  if (variant.selectedOptions) {
+    const firstSelectedOption = variant.selectedOptions[0];
+    if (!firstSelectedOption?.name) {
+      validParam = logErrorAndReturnFalse(`${key}.selectedOptions.name`);
+    }
+    if (!firstSelectedOption?.value) {
+      validParam = logErrorAndReturnFalse(`${key}.selectedOptions.value`);
+    }
+  } else {
+    validParam = logErrorAndReturnFalse(`${key}.selectedOptions`);
+  }
+
+  return validParam;
+}
+
+/**
+ * Finds all the variants provided by adjacentVariants, options.optionValues.firstAvailableVariant,
+ * and selectedOrFirstAvailableVariant and return them in a single array
+ */
+export function getAdjacentAndFirstAvailableVariants(
+  product: RecursivePartial<Product>,
+): ProductVariant[] {
+  // Checks for valid product input
+  const checkedProduct = checkProductParam(product);
```

Comment:
Why does sometimes `checkProductParm` "checkAll" vs other times not?

Reply:
I want to reuse the `checkProductParam` with the `getAdjacentAndFirstAvailableVariants`. 

The purpose of `getAdjacentAndFirstAvailableVariants` is to return a list of variants found in `options.optionValues.firstAvailableVariant`, `adjacentVariants`, and `selectedOrFirstAvailableVariant`. The `getAdjacentAndFirstAvailableVariants` doesn't require other fields to be present like `encodedVariantExistence`, and `encodedVariantAvailability`